mailbox/api/src/main/java/org/apache/james/mailbox/MailboxListener.java

```diff
@@ -314,65 +324,96 @@ public MetaDataHoldingEvent(MailboxSession.SessionId sessionId, User user, Mailb
+
+        public Expunged(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId, Map<MessageUid, MessageMetaData> uids) {
             super(sessionId, user, path, mailboxId);
+            this.uids = uids;
+        }
+
+        @Override
+        public List<MessageUid> getUids() {
+            return ImmutableList.copyOf(uids.keySet());
         }
 
         /**
          * Return the flags which were set for the added message
          *
          * @return flags
+         *
          */
         @Override
-        public abstract MessageMetaData getMetaData(MessageUid uid);
+        public MessageMetaData getMetaData(MessageUid uid) {
+            return uids.get(uid);
+        }
     }
 
     /**
      * A mailbox event related to updated flags
      */
-    abstract class FlagsUpdated extends MessageEvent {
+     class FlagsUpdated extends MessageEvent {
 
         /**
          *
          */
         private static final long serialVersionUID = 1L;
 
-        public FlagsUpdated(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId) {
+        private final List<MessageUid> uids;
+        private final List<UpdatedFlags> uFlags;
+
+        public FlagsUpdated(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId, List<MessageUid> uids, List<UpdatedFlags> uFlags) {
             super(sessionId, user, path, mailboxId);
+            this.uids = ImmutableList.copyOf(uids);
+            this.uFlags = ImmutableList.copyOf(uFlags);
         }
 
-        public abstract List<UpdatedFlags> getUpdatedFlags();
+        @Override
+        public List<MessageUid> getUids() {
+            return uids;
+        }
+
+        public List<UpdatedFlags> getUpdatedFlags() {
+            return uFlags;
+        }
     }
 
     /**
      * A mailbox event related to added message
      */
-    abstract class Added extends MetaDataHoldingEvent {
+    class Added extends MetaDataHoldingEvent {
 
         /**
          *
          */
         private static final long serialVersionUID = 1L;
 
-        public Added(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId) {
+        private final Map<MessageUid, MessageMetaData> added;
+
+        public Added(MailboxSession.SessionId sessionId, User user, MailboxPath path, MailboxId mailboxId, SortedMap<MessageUid, MessageMetaData> uids) {
             super(sessionId, user, path, mailboxId);
+            this.added = ImmutableMap.copyOf(uids);
         }
 
         /**
          * Return the flags which were set for the added message
          * 
          * @return flags
          */
+        public MessageMetaData getMetaData(MessageUid uid) {
+            return added.get(uid);
+        }
+
         @Override
-        public abstract MessageMetaData getMetaData(MessageUid uid);
-        
+        public List<MessageUid> getUids() {
+            return ImmutableList.copyOf(added.keySet());
```

Comment:
same comment about the place you do the copy

Reply:
It is a conversion to `List` and not a copy for immutability

Reply:
could be changed to Set, Collection or Iterable in MessageEvent IMO